modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,698 @@
+                GAPI_Assert(!m_oak_nodes.at(nh).inputs.empty());
+                GAPI_Assert(!m_oak_nodes.at(nh).outputs.empty());
+
+                GOAKContext ctx(m_pipeline, m_camera_size, in_ctx_args, m_oak_nodes.at(nh).outputs);
+                m_oak_nodes.at(nh).node = u.k.m_put_f(ctx, m_in_queues);
+                GAPI_Assert(m_oak_nodes.at(nh).node != nullptr);
+            }
+        }
+
+        // Check that all inputs and outputs are properly filled after constructing kernels
+        // to then link it together
+        for (const auto& node : m_oak_nodes) {
+            auto el = node.second;
+            if (std::any_of(el.inputs.cbegin(), el.inputs.cend(),
+                            [](ExtractTypeHelper::InputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI input are not set");
+            }
+            if (std::any_of(el.outputs.cbegin(), el.outputs.cend(),
+                            [](ExtractTypeHelper::OutputPtr ptr) {
+                    return ptr == nullptr;
+                })) {
+                GAPI_Assert(false && "DAI outputs are not set");
+            }
+        }
+
+        // Prepare nodes for linking
+        std::vector<ade::NodeHandle> in_nodes;
+        std::vector<ade::NodeHandle> out_nodes;
+        std::vector<ade::NodeHandle> inter_nodes;
+
+        for (const auto& node : m_oak_nodes) {
+            auto nh = node.first;
+            // Fill input op nodes
+            for (const auto& d : ins_data) {
+                for (const auto& indata : nh.get()->inNodes()) {
+                    auto rc = m_gm.metadata(indata).get<cv::gimpl::Data>().rc;
+                    if (rc == d.rc) {
+                        in_nodes.push_back(nh);
+                    }
+                }
+            }
+            // Fill output op nodes
+            for (const auto& d : outs_data) {
+                for (const auto& outdata : nh.get()->outNodes()) {
+                    auto rc = m_gm.metadata(outdata).get<cv::gimpl::Data>().rc;
+                    if (rc == d.rc) {
+                        out_nodes.push_back(nh);
+                    }
+                }
+            }
+            // Fill internal op nodes
+            if (std::find(in_nodes.begin(), in_nodes.end(), nh) == in_nodes.end() &&
+                std::find(out_nodes.begin(), out_nodes.end(), nh) == out_nodes.end()) {
+                inter_nodes.push_back(nh);
+            }
+        }
+
+        // Properly link all nodes
+        // 1. Link input nodes to camera
+        for (const auto& nh : in_nodes) {
+            GAPI_Assert(m_oak_nodes.at(nh).inputs.size() == 1);
+            // FIXME: covert other camera outputs
+            m_camera_input->video.link(*(m_oak_nodes.at(nh).inputs[0]));
+        }
+
+        // 2. Link output nodes to XLinkOut nodes
+        size_t out_counter = 0;
+        for (const auto& nh : out_nodes) {
```

Comment:
there's no strict defined order in `out_nodes` right?

Reply:
It's aligned with the backend's `out_data` and is being used in that aligned way